Corrected logic in the SchemaBase.Status

MarkGaze opened the pull request:
The current logic is checking that the Parent is not null but then the ORs allow the code to continue, I'm assuming that the last three statements were meant to be grouped. Can you check the validity of this @igitur? I ran it and it fixed #73 for me.

igitur commented:
I want to reproduce the problem on my side before accepting any PRs.

MarkGaze commented:
I've added a unit test to prove the fix, if you change back to the old code you can recreate the issue I had on my machine which is fixed by the code change

igitur commented:
I'm not convinced by this PR. One last time... please attach schema creation scripts for objects that reproduces this problem.

MarkGaze commented:
I don't know which of the thousands of objects that are in my databases are creating the issue, and I haven't got the time to spend searching for them otherwise I would gladly send them.

It seems fairly clear that the logic in that **one line of code** I changed in the original commit is incorrect, as it allows the execution of the code inside it regardless of whether the Parent is null or not and that code contains usages of Parent.Status on all other branches of code.

The unit test that I added proves this further.

So, either the fix I made (which fixed both my problem and the unit test I wrote that recreated the issue) is correct or the if statement needs rethinking.

I'm evaluating this tool for use in my workplace, which has a lot of large databases so I expected problems but the open nature of the codebase is a major benefit to me and I'm more than happy to contribute to it but having to prove each fix with scripts is going to be a problem which is why I added the unit test that performs the same function, if there are other critical paths that would use this code then I'm happy to discuss and add unit tests for them too.

igitur commented:
You don't have to go through thousands of objects. The error message in #73 states that the culprit object is `DocumentInsert`.

`Error Comparing Databases (while Reading Text Objects: [DocumentInsert])`.

Unfortunately the current unit test coverage is extermely poor, so adding a unit test to substantiate your PR is not enough. It is very possible that this PR could break other areas while passing the unit test that you added. Therefore I'm not going to blindly accept code changes without understanding the actual cause of the problem.

MarkGaze commented:
Except (and this is my fault, because I removed the comment with the files in which had this information) the file changed to a different one after I logged the issue. I'll have another look.

igitur commented on OpenDBDiff.Schema/Model/SchemaBase.cs:
```diff
@@ -229,7 +229,7 @@ public virtual ObjectStatus Status
 
                 // Si el estado de la tabla era el original, lo cambia, sino deja el actual estado.
-                if (Parent != null && Parent.Status == ObjectStatus.Original || value == ObjectStatus.Rebuild || value == ObjectStatus.RebuildDependencies)
+                if (Parent != null && (Parent.Status == ObjectStatus.Original || value == ObjectStatus.Rebuild || value == ObjectStatus.RebuildDependencies))
```
I agree with this fix, but I'd like to rework the logic in this property a bit. But I still want you to get the credit. Would you mind changing the entire `SchemaBase.Status` property to:

```c#
        public virtual ObjectStatus Status
        {
            get { return status; }
            set
            {
                if (status != ObjectStatus.Rebuild && status != ObjectStatus.RebuildDependencies)
                    status = value;

                if (Parent == null) return;

                // Si el estado de la tabla era el original, lo cambia, sino deja el actual estado.
                // If the state of the table was the original, it changes it, but leaves the current state. (Google translated)
                if (Parent.Status == ObjectStatus.Original
                    || value == ObjectStatus.Rebuild
                    || value == ObjectStatus.RebuildDependencies)
                {
                    switch (value)
                    {
                        case ObjectStatus.RebuildDependencies:
                        case ObjectStatus.Rebuild:
                            Parent.Status = value;
                            break;

                        case ObjectStatus.Original:
                            break;

                        default:
                            Parent.Status = ObjectStatus.Alter;
                            break;
                    }
                }
            }
        }
```

igitur commented on OpenDBDiff.sln:
```diff
@@ -27,6 +27,8 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "OpenDBDiff.Schema.SQLServer
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "OpenDBDiff.Schema.SQLServer.GeneratesTests", "OpenDBDiff.Schema.SQLServer.GeneratesTests\OpenDBDiff.Schema.SQLServer.GeneratesTests.csproj", "{94F5C91A-6268-41E9-A15E-FA5451BB2B6F}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "OpenDBDiff.SchemaTests", "OpenDBDiff.SchemaTests\OpenDBDiff.SchemaTests.csproj", "{9CBBBA4C-AA7E-41DE-B277-2F5CE7D31ACC}"
```
Revert, as discussed.

MarkGaze commented:
@igitur I think we should be good now!

jspraul commented:
Much thanks to @MarkGaze for doing the exra work to polish this up per @igitur's guidance.

igitur commented:
Thanks for the contribution.
